### PR TITLE
Resolved problems with plugins loading ordering and added type safe ordering function

### PR DESCRIPTION
**Pull request by BarkingBad**

### CODE REVIEW

File: core/src/main/kotlin/plugability/DokkaContext.kt [hunk 1]
```diff
@@ -68,6 +69,49 @@ private class DokkaContextConfigurationImpl(
...
+        VISITING,
+        VISITED;
+    }
+
+    internal val verticesWithState = mutableMapOf<Extension<*>, State>()
+
+    internal val adjacencyList: MutableMap<Extension<*>, MutableList<Extension<*>>> = mutableMapOf()
+
+    private fun topologicalSort() {
+
```
Comment (sellophane) [inline core/src/main/kotlin/plugability/DokkaContext.kt]: Why so many whitespaces?

File: core/src/main/kotlin/plugability/DokkaContext.kt [hunk 2]
```diff
@@ -68,6 +69,49 @@ private class DokkaContextConfigurationImpl(
...
+            val state = verticesWithState[n]
+
+            if(state == State.VISITED)
+                return
+            if(state == State.VISITING)
+                throw Error("Detected cycle in plugins graph")
+
+            verticesWithState[n] = State.VISITING
+
+            for(m in adjacencyList[n]!!)
```
Comment (sellophane) [inline core/src/main/kotlin/plugability/DokkaContext.kt]: adjacencyList[n]?.forEach { visit(it) }

File: core/src/main/kotlin/plugability/DokkaContext.kt (hunk 1 above)
Comment (BarkingBad) [reply core/src/main/kotlin/plugability/DokkaContext.kt]: Hard to tell, was more readable while writing

File: core/src/main/kotlin/plugability/DokkaContext.kt (hunk 2 above)
Comment (BarkingBad) [reply core/src/main/kotlin/plugability/DokkaContext.kt]: Will change

File: core/src/main/kotlin/plugability/DokkaContext.kt (hunk 1 above)
Comment (BarkingBad) [reply core/src/main/kotlin/plugability/DokkaContext.kt]: Will change

**Pull request merged**